### PR TITLE
Reattempt parallel consistent mortar mechanical contact

### PR DESCRIPTION
**Pull request by lindsayad**
Previous technical discussion about the content of this PR is at #19246. Should work now that we've fixed leaks from TIMPI

### CODE REVIEW

**Review by cticenhour — DISMISSED** [review]
Please add a libmesh update summary to `modules/doc/content/newsletter/2021_11.md`

**Comment by lindsayad** [comment]
Alright, we passed Valgrind contact!

**Comment by lindsayad** [comment]
Ok @fdkong sets `skip_partitioning(true)` in his distributed mesh generator. But when we recover `_skip_all_partitioning` is set to `false`. We could debate whether we should write out that data member to checkpoint so that the flag is the same upon recover as it is before recover, but that's not the main point here ...

When we have both distributed and displaced mesh, I think that geometric proxy relationship managers simply won't work when repartitioning is allowed. When the reference mesh is repartitioned it will redistribute its elements. Then when we call repartition on the displaced mesh, we will be querying its ghosting functors, one of which will be the geometric proxy relationship manager. This will look over at the reference mesh and run its ghosting functors with `reference_mesh.active_local_elements_begin()` and `reference_mesh.active_local_elements_end()` as calling args and return a set of elements. What is happening in the failing test is that the returned set has elements that do not yet exist on the displaced mesh (as ascertained through checking unique id) because we *have not yet* redistributed the displaced mesh; we are in fact querying the ghosting functors to help us determine what elements to send/receive in our redistribution. And because this reference element does not yet have its displaced mesh element version on the current process, we fail the `unique_id_to_elem_map` find assertion.

I think my conclusion here is that proxy relationship managers are not the solution for maintaining synchronicity between meshes. I'm going to drop that code from this PR. This kind of stuff really makes me want to get https://github.com/libMesh/libmesh/issues/2661 going @roystgnr

**Comment by fdkong** [comment]
>I think my conclusion here is that proxy relationship managers are not the solution for maintaining synchronicity between meshes. I'm going to drop that code from this PR. This kind of stuff really makes me want to get libMesh/libmesh#2661 going @roystgnr

I agreed. Sync between the reference and displaced meshes is extremely painful, especially when we have distributed mesh, mesh adaptivity, etc.

File: modules/contact/test/tests/bouncing-block-contact/mixed-weighted-gap-swapped.i
Comment (recuero) [inline modules/contact/test/tests/bouncing-block-contact/mixed-weighted-gap-swapped.i:69]: I'll try to remember to expose this parameter through the action next time we update it.

File: modules/contact/test/tests/bouncing-block-contact/variational-frictional.i
Comment (recuero) [inline modules/contact/test/tests/bouncing-block-contact/variational-frictional.i:179]: I believe there is a way (IIRC) to set this parameter within the code. If that's possible, we may want to set it within the mortar constraints --eventually?

**Review by recuero — DISMISSED** [review]
Looks great to me.

File: modules/contact/test/tests/bouncing-block-contact/variational-frictional.i
Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/variational-frictional.i:179]: I think it may actually be unnecessary to set this to false for mortar. Or if it is necessary I don’t understand why haha. This should not be stateful like some node-face code, eg our residual should be entirely determined by the current solution. Like I know that if a node ping pongs a couple times in the node-face case we lock it in place which is an action independent of the current solution.

Comment (recuero) [reply modules/contact/test/tests/bouncing-block-contact/variational-frictional.i:179]: I believe we started setting that flag to false when we were using `NormalNodalLMMechanicalContact` stuff (https://github.com/idaholab/moose/issues/16821) which partially relied on node-face machinery. I have been carried that setting even with the latest full-mortar contact objects, probably unnecessarily as you point out.

Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/variational-frictional.i]: Alright I removed this flag, and things passed locally. Hopefully it will pass on CIVET too

**Review by recuero — APPROVED** [review]

**Comment by lindsayad** [comment]
@cticenhour the libmesh update has been removed from this PR so hopefully your concern is addressed

**Comment by cticenhour** [comment]
@lindsayad I have dismissed my previous review. However, as I am on PL, I feel I should defer to someone else on the CCB to make the final approval and merge who is more familiar with the content.

File: framework/src/constraints/MortarConstraintBase.C
Comment (fdkong) [inline framework/src/constraints/MortarConstraintBase.C:191]: What `sn` and `vn` stand for?

File: modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C
Comment (fdkong) [inline modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:175]: IF this has no sparse AD, will the gaps be correct as well? If not, we might add some messages to inform users

**Review by fdkong — CHANGES_REQUESTED** [review]
It looks good to me.

Just minor queries

File: framework/src/constraints/MortarConstraintBase.C
Comment (lindsayad) [reply framework/src/constraints/MortarConstraintBase.C:191]: Haha I can change the variable names to sys_number and var_number

File: modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:175]: We error out in the constructor if the user is using nonsparse

Comment (fdkong) [reply modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:175]: Good

**Review by fdkong — APPROVED** [review]

**Pull request merged**
